### PR TITLE
coprocessor/dag: use expression in dag

### PR DESCRIPTION
**Pull request by AndreMouche**
@hicqu @BusyJay @hanfei1991 PTAL

### CODE REVIEW

**Comment by AndreMouche** [comment]
@BusyJay @hicqu @hanfei1991 PTAL

File: src/coprocessor/dag/executor/mod.rs [hunk 1]
```diff
@@ -104,35 +104,34 @@ pub trait Executor {
...
-    eval: &mut Evaluator,
     ctx: &EvalContext,
     values: &RowColsDict,
     columns: Rc<Vec<ColumnInfo>>,
     offsets: &[usize],
     h: i64,
-) -> Result<()> {
+) -> Result<Vec<Datum>> {
+    let mut res = Vec::with_capacity(columns.len());
+    res.resize(columns.len(), Datum::Null);
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/mod.rs]: Why not use `let mut res = vec![Datum::Null; columns.len()]`?

File: src/coprocessor/dag/executor/aggregation.rs [hunk 2]
```diff
@@ -15,22 +15,77 @@ use std::rc::Rc;
...
+    fn build(ctx: &EvalContext, mut expr: Expr) -> Result<AggrFuncExpr> {
+        let args = box_try!(Expression::batch_build(
+            expr.take_children().into_vec(),
+            ctx
+        ));
+        let tp = expr.get_tp();
+        Ok(AggrFuncExpr { args: args, tp: tp })
+    }
+
+    fn eval_args(&self, ctx: &EvalContext, row: &[Datum]) -> Result<Vec<Datum>> {
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: s/eval_args/eval/

File: src/coprocessor/dag/executor/aggregation.rs [hunk 3]
```diff
@@ -71,14 +126,14 @@ impl<'a> AggregationExecutor<'a> {
         })
     }
 
-    fn get_group_key(&mut self, eval: &mut Evaluator) -> Result<Vec<u8>> {
+    fn get_group_key(&mut self, row: &[Datum]) -> Result<Vec<u8>> {
```
Comment (hicqu) [inline src/coprocessor/dag/executor/aggregation.rs]: `self` doesn't need to be `mut`.

**Comment by hicqu** [comment]
rest LGTM.

**Comment by AndreMouche** [comment]
@BusyJay @hanfei1991 PTAL

File: src/coprocessor/dag/executor/aggregation.rs [hunk 4]
```diff
@@ -15,22 +15,77 @@ use std::rc::Rc;
...
 use coprocessor::select::aggregate::{self, AggrFunc};
-use coprocessor::select::xeval::{EvalContext, Evaluator};
+use coprocessor::select::xeval::EvalContext;
+use coprocessor::dag::expr::Expression;
 use coprocessor::metrics::*;
 use coprocessor::Result;
 
 use super::{inflate_with_col_for_dag, Executor, ExprColumnRefVisitor, Row};
 
+struct AggrFuncExpr {
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: Can this be merged with `AggrFunc`?

Comment (AndreMouche) [reply src/coprocessor/dag/executor/aggregation.rs]: address comments

File: src/coprocessor/dag/executor/aggregation.rs [hunk 5]
```diff
@@ -15,22 +15,77 @@ use std::rc::Rc;
...
+    }
+
+    fn eval_args(&self, ctx: &EvalContext, row: &[Datum]) -> Result<Vec<Datum>> {
+        let res: Vec<Datum> = box_try!(self.args.iter().map(|v| v.eval(ctx, row)).collect());
+        Ok(res)
+    }
+}
+
+impl AggrFunc {
+    fn build_with_expr(
```
Comment (BusyJay) [inline src/coprocessor/dag/executor/aggregation.rs]: This function is unnecessary.

File: src/coprocessor/dag/expr/mod.rs [hunk 6]
```diff
@@ -294,6 +285,15 @@ impl Expression {
         }
     }
 
+    pub fn batch_build(exprs: Vec<Expr>, ctx: &StatementContext) -> Result<Vec<Self>> {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: Put `ctx` first.

**Comment by BusyJay** [comment]
CI fail.

**Pull request merged**